### ISSUE TITLE
`require` fails when requiring a macro-utils file from other macro files (Fennel Issue #406)

### ISSUE THREAD

**Issue by datwaft** [Olical/conjure#286]
It seems that because Conjure disables the compiler sandbox it is also affected by https://github.com/bakpakin/Fennel/issues/406.

Inside the issue is how to reproduce the issue.

![image](https://user-images.githubusercontent.com/37723586/146661279-2715e47e-81d9-46d0-a31e-ebead8f67476.png)

Is there a way to activate the compiler sandbox so this issue doesn't happen?

I opened this issue mainly for tracking, but a workaround would be wonderful.

**Comment by Olical** [comment]
So this is `compilerEnv: _G` doing this? This is required for the Aniseed module macros to work as far as I'm aware 🤔 I'll have a tinker though.

**Comment by Olical** [comment]
Yeah, if I don't set that I can't access `package.loaded` in the module macros 🤔

**Comment by datwaft** [comment]
> So this is `compilerEnv: _G` doing this?

Yes, you can see https://github.com/rktjmp/hotpot.nvim/issues/48 for more specific information. Its a  combination of another compiler option together with `compilerEnv = _G`.

**Comment by datwaft** [comment]
https://github.com/bakpakin/Fennel/issues/406 is fixed now 🎉

**Comment by datwaft** [comment]
Is there a planned date for Conjure to use Fennel v1.1.0?

As you can see this issue is still happening on the latest version of Conjure:

![image](https://user-images.githubusercontent.com/37723586/164136464-c3716834-a0c5-44ad-863d-712c4e345d6f.png)

But the bug was fixed on the release of Fennel v1.1.0 ([here](https://fennel-lang.org/v1.1.0/) are the release notes).

> - Fix a bug where disabling the compiler sandbox broke module require scope

**Comment by Olical** [comment]
Oh nice! I'll get that updated soon! Thank you for the heads up!

**Comment by Olical** [comment]
This is merged and released now, will close assuming it's all good now. If not, something else might be wrong!

**Comment by datwaft** [comment]
![image](https://user-images.githubusercontent.com/37723586/166257325-cb838785-c945-4a50-97bf-8a5812c45e35.png)

It seems that the bug is still reproduced.

On both Hotpot and raw Fennel the bug is no longer reproducible, so it is weird that it is reproduced on Conjure.

**Comment by Olical** [comment]
Hmm, okay, I'll have to try and reproduce it. You're sure you had the latest commit from this morning?

```
commit 842c81892648de759e639ad2d395757b98be06d5 (tag: v4.33.2, origin/master, origin/HEAD, master)
Author: Oliver Caldwell <ollie@oli.me.uk>
Date:   Fri Apr 29 09:32:27 2022 +0100
```

I've run into a few issues where people thought they'd updated but their plugin manager hadn't done it for some reason.

**Comment by Olical** [comment]
Also trying to understand the issue: If you try to require macros from another macro file it fails to load the helpers which breaks the parent file?

**Comment by Olical** [comment]
I can't reproduce this, I used `(import-macros...` to import a file then in there I pulled in another macro file and I can reference them no problem 🤔  which lends to my "something didn't update fully or properly" theory, but I'm very open to being proven wrong!

**Comment by Olical** [comment]
The other thing it could be is that your macro helpers file isn't on the path like the other macro modules are for some reason.

**Comment by datwaft** [comment]
> Hmm, okay, I'll have to try and reproduce it. You're sure you had the latest commit from this morning?

As you can see in the screenshot, yes:

![image](https://user-images.githubusercontent.com/37723586/166268045-cbe9a298-c358-4412-8845-bbebc7798165.png)

> Also trying to understand the issue: If you try to require macros from another macro file it fails to load the helpers which breaks the parent file?

Not exactly, its if I try to require helper functions from a macro file.

I have a file called `helpers.fnl` which contains some helper functions like `fn?`, which checks if the input is a list and if the first element is the symbol `hashfn` or `fn`, when requiring this file from a macro file the bug is produced.

Here is an example of the code: https://github.com/datwaft/fennel-issue_macro_require

**Comment by datwaft** [comment]
Here is the important snippet:

```fennel
(local {: fn?} (require :macro-utils))

(lambda if-fn-hello-else-bye [x]
  (if
    (fn? x) `(print "Hello")
    `(print "Bye")))

{: if-fn-hello-else-bye}
```

**Comment by Olical** [comment]
Got it, I understand now, will try to repro and work it out.

**Comment by datwaft** [comment]
Are there any updates?

I am still having this issue and it blocks me for using Conjure for testing my Neovim Fennel configuration.

<img width="896" alt="image" src="https://user-images.githubusercontent.com/37723586/185731650-26da3a43-6b62-49dc-ab69-4144481b51e1.png">

As I said before the issue should have been solved in Fennel, so it's probably something related to Aniseed or Conjure.

Thank you for your help, btw!

**Comment by Olical** [comment]
https://github.com/Olical/aniseed/issues/128 I think this is just this issue too. Haven't fixed it yet, sorry 😢 I keep working on other things and have been really busy recently. Will try to get this fixed very soon. If anyone wants to look into it, I think we just need to set the fennel path to the same as the lua path or something in the Aniseed compiler module.

**Comment by datwaft** [comment]
Hello!

It has been a few months so I thought I would ask if there has been any progress with this issue.

As you can see here I can still reproduce it:

<img width="1552" alt="image" src="https://user-images.githubusercontent.com/37723586/210267445-c5326767-743e-48fb-8150-114cd0d22a47.png">

**Comment by Olical** [comment]
I just spent another 3 hours trying to get to the bottom of this and still can't work it out, I'm sorry. I just don't know how to get the fennel compiler to do what I want. I can however get this working just fine if I use `import-macros` to load my macro util file, that works perfectly with static compilation and dynamic eval through Conjure.

I really need some help with this, if anyone can spot what args I can set or change to get the fennel compiler to treat requires inside macro files as further macro requires, please let me know.

Relevant lines are here https://github.com/Olical/aniseed/blob/a7445c340fb7a0529f3c413eb99d3f8d29f50ba2/fnl/aniseed/compile.fnl#L26-L42

Maybe it's because I use `compileString` instead of some file based functions? I've also made it so (on the dev branch) anything under a `macro` or `macros` directory is considered a macro and won't be statically compiled, hopefully this makes macro file organisation easier.

This test bed works for me, for context:

```fnl
;; main.fnl
(import-macros {: if-fn-hello-else-bye} :macros.user)

(if-fn-hello-else-bye :hey)


;; macros/user.fnl
(import-macros {: is-fn} :macros.utils)

(lambda if-fn-hello-else-bye [x]
  (if
    (is-fn x) `(print "Hello")
    `(print "Bye")))

{: if-fn-hello-else-bye}


;; macros/utils.fnl
{:is-fn (fn [x]
                (= `fn (. x 1)))}
```

It's just the ability to use a regular require inside a macro file and for the compiler to know to load that as a macro as well that doesn't work. I've updated the Fennel compiler, maybe there's an option I need to set to get that working? Possibly to do with the options I provide to the compiler?

```fnl
            (fnl.compileString
              (a.merge! {:allowedGlobals false
                         :compilerEnv _G} opts))
```

**Comment by datwaft** [comment]
Thanks for your effort!

Maybe @rktjmp can help as Hotpot doesn't have any problem with my use case after https://github.com/bakpakin/Fennel/issues/406 was merged.

Sadly I don't have much experience tinkering with Fennel's internals so I don't have proper idea about how to fix this.

Maybe it's an upstream issue if `fennel.compileString` has the issue and https://github.com/bakpakin/Fennel/issues/406 was only fixed for `fennel.load` or something like that.

**Fun fact:** when using stdio with Fennel it works without any problem:

<img width="2672" alt="image" src="https://user-images.githubusercontent.com/37723586/211163866-420f6ee2-54ca-4fc0-89f9-3cc857758b6b.png">

As I don't need Aniseed maybe a workaround could be to use the stdio and somehow add to its path Neovim's API.

---

> This test bed works for me, for context:

Btw, that test bed would always output `Bye` as the symbol `x` is what is passed to `is-fn`. That is why it's important to be able to use `require` instead of `import-macros`.

<img width="2628" alt="image" src="https://user-images.githubusercontent.com/37723586/211164179-84a392da-5282-4061-8f9e-deafae17f07e.png">
